File: 图/最短路径/Floyd算法/Floyd.java
public class Floyd {
    public final static int INFINITY = Integer.MAX_VALUE / 2; //防止溢出

    public static int floyd(int[] vertexes, int[][] arches,
            int i, int j, int k, int[][] D, int[][] P) {
        if (k == 0)
            return arches[i][j];
        if (i == j)
            return D[i][j] = 0;
        if (D[i][j] != INFINITY)
            return D[i][j];

        // 选择k进行中转
        int lengthK = floyd(vertexes, arches, i, k, k - 1, D, P) +
            floyd(vertexes, arches, k, j, k - 1, D, P);
        // 不选择k进行中转
        int lengthWithoutK = floyd(vertexes, arches, i, j, k - 1, D, P);
        D[i][j] = lengthWithoutK;
        if (lengthK < lengthWithoutK) {
            D[i][j] = lengthK;
            P[i][j] = k;
        }
        return D[i][j];
    }

    public static void main(String[] args) {
        // 顶点0是用来占位的
        int[] vertexes = new int[]{0, 1, 2, 3, 4, 5};
        int[][] arches = new int[][]{
            //           0         1         2         3         4         5
            new int[]{INFINITY, INFINITY, INFINITY, INFINITY, INFINITY, INFINITY}, // 0
            new int[]{INFINITY, INFINITY,        3,        8, INFINITY,       -4}, // 1
            new int[]{INFINITY, INFINITY, INFINITY, INFINITY,        1,        7}, // 2
            new int[]{INFINITY, INFINITY,        4, INFINITY, INFINITY, INFINITY}, // 3
            new int[]{INFINITY,        2, INFINITY,       95, INFINITY, INFINITY}, // 4
            new int[]{INFINITY, INFINITY, INFINITY, INFINITY,        6, INFINITY}, // 5
        };
        int size = vertexes.length;

        int[][] D = new int[size][size]; // D[i][j]表示从顶点i 到 顶点j的最短路径
        int[][] P = new int[size][size]; // P[i][j]表示从顶点i 到 顶点j要经过的顶点
        for (int i = 0; i < size; i++)
            for (int j = 0; j < size; j++) {
                D[i][j] = INFINITY;
                P[i][j] = j;
            }

        for (int i = 1; i < size; i++)
            for (int j = 1; j < size; j++)
                floyd(vertexes, arches, i, j, size-1, D, P);

        for (int i = 1; i < size; i++)
            for (int j = 1; j < size; j++)
                System.out.println(i + " ---> " + j + " = "
                    + D[i][j] + ", 经过" + P[i][j]);
    }
}

File: 图/最短路径/Floyd算法/floyd.py
# coding: utf8


class Graph(object):
    """
    图的数组矩阵表示法
    """
    def __init__(self):
        self._vertexes = []
        self._arches = []

    def add_vertex(self, element):
        self._vertexes.append(element)
        for arch in self._arches:
            arch.append(None)
        self._arches.append([None] * len(self._vertexes))

    def get_vertex_count(self):
        return len(self._vertexes)

    def add_arch(self, tail, head, weight):
        self._arches[tail][head] = weight

    def get_weight(self, tail, head):
        return self._arches[tail][head]


def floyd(graph):
    """
    Floyd 算法实现
    """
    import sys

    INFINITY = sys.maxint / 2.0
    vertex_count = graph.get_vertex_count()

    # 初始化辅助数组 d[][]、p[][]
    d = []
    p = []

    # 初始时，任意两个顶点之间的最短路径是 INFINITY
    for i in range(vertex_count):
        d.append([])
        for _ in range(vertex_count):
            d[i].append(INFINITY)

    # 初始时，p[i][j] = j
    for i in range(vertex_count):
        p.append([])
        for j in range(vertex_count):
            p[i].append(j)

    def _floyd(i, j, k):
        if k == -1:
            return graph.get_weight(i, j) or INFINITY
        if i == j:
            d[i][j] = 0
            return 0
        if d[i][j] != INFINITY:
            return d[i][j]

        with_k = _floyd(i, k, k - 1) + _floyd(k, j, k - 1)
        without_k = _floyd(i, j, k - 1)

        if with_k < without_k:
            d[i][j] = with_k
            p[i][j] = k
            return with_k

        d[i][j] = without_k
        return without_k

    for i in range(vertex_count):
        for j in range(vertex_count):
            _floyd(i, j, vertex_count - 1)

    return d, p


def test():
    # http://images.timd.cn/data-structure/floyd-example.png
    g = Graph()
    for ind in range(1, 5):
        g.add_vertex(ind)
    for arch in [(0, 1, 2), (0, 2, 6), (0, 3, 4),
                 (1, 2, 3), (2, 0, 7), (2, 3, 1),
                 (3, 0, 5), (3, 2, 12)]:
        g.add_arch(*arch)
    d, p = floyd(g)
    print(d)
    print(p)


if __name__ == "__main__":
    test()
